### PR TITLE
Handle NULL pointers for string args

### PR DESCRIPTION
**Pull request by andrewstrohman**
This change lays the ground work for addressing https://github.com/cilium/tetragon/issues/4579. After I created the error arg type with [this change](https://github.com/cilium/tetragon/pull/4327), the problem described in that issue mutated such that we send events, but they are corrupted.

While looking into that problem, I found another issue with https://github.com/cilium/tetragon/pull/4327 where returnArgs can be overwritten when the first configured arg read results in an error arg.

Userspace probes have different string reading logic than kprobes. While events were not dropped in the uprobe case when NULL pointers are passed, empty strings were being reported in the event. So, I made uprobes consistent with kprobes in this regard, by returning an error arg when a NULL pointer is passed. This way, we can distinguish between empty strings and NULL pointers.

While creating the uprobe test I noticed a bug where I was masking against a macro that was not actually a mask. I corrected this as well.

I wanted to make the uprobe tests policytests, but I couldn't because there is no way to determine kfunc availability for policy tests yet.

### CODE REVIEW

File: bpf/process/types/basic.h [hunk 1]
```diff
@@ -365,8 +365,8 @@ FUNC_INLINE long copy_strings(char *args, char *arg, int max_size)
 	// So add one to the length to allow for it. This should
 	// result in us honouring our max_size correctly.
 	size = probe_read_str(&args[4], max_size + 1, arg);
-	if (size <= 0)
-		return invalid_ty;
+	if (size < 0)
+		return size;
```
Comment (kkourt) [inline bpf/process/types/basic.h:369]: This changes a `<=` check to a `<` check, and below we do `size--`. Is there  a chance for an overflow?

File: bpf/process/generic_calls.h [hunk 2]
```diff
@@ -657,13 +667,6 @@ generic_process_event(void *ctx, struct bpf_map_def *tailcals, int process)
 		errv = generic_read_arg(ctx, index, total, tailcals, process);
 		if (errv > 0)
 			total += errv;
-		/* Follow filter lookup failed so lets abort the event.
-		 * From high-level this is a filter and should be in the
-		 * filter block, but its just easier to do here so lets
-		 * do it where it makes most sense.
-		 */
-		if (errv < 0)
-			return filter_args_reject(e->func_id);
```
Comment (kkourt) [inline bpf/process/generic_calls.h:666]: Why is this removed? Can you please add details in the commit message?

File: bpf/process/generic_calls.h [hunk 3]
```diff
@@ -467,22 +467,32 @@ read_arg(void *ctx, int index, int type, long orig_off, unsigned long arg, int a
...
-		return __read_arg_1(ctx, type, orig_off, arg, argm, args);
-	if (process == __READ_ARG_2)
-		return __read_arg_2(ctx, type, orig_off, arg, argm, args);
-
-	/* .. and the rest of the world */
-	if (is_read_arg_1(type))
-		return __read_arg_1(ctx, type, orig_off, arg, argm, args);
-	else
-		return __read_arg_2(ctx, type, orig_off, arg, argm, args);
+	if (path_arg) {
```
Comment (kkourt) [inline bpf/process/generic_calls.h:474]: Why is path_arg special? Can you add a commit or the specifics in the commit message?

File: pkg/sensors/tracing/tracepoint_test.go [hunk 4]
```diff
@@ -479,7 +479,7 @@ func TestLoadTracepointSensor(t *testing.T) {
 		sensorMaps = append(sensorMaps, tus.SensorMap{Name: "tp_calls", Progs: []uint{0, 1, 2, 3, 4, 6, 7}})
 
 		// all but generic_tracepoint_event,generic_tracepoint_filter
-		sensorMaps = append(sensorMaps, tus.SensorMap{Name: "retprobe_map", Progs: []uint{1, 2, 6, 7}})
+		sensorMaps = append(sensorMaps, tus.SensorMap{Name: "retprobe_map", Progs: []uint{1, 6, 7}})
```
Comment (kkourt) [inline pkg/sensors/tracing/tracepoint_test.go:482]: > This is the problematic code path that caused dropped events, which the previous commit fixed.

Thanks for fixing this, but it's unclear to me what the issue was. Can you add some details in the commit message?

File: pkg/sensors/tracing/args_linux.go [hunk 5]
```diff
@@ -112,7 +112,11 @@ func getArgStatus(r *bytes.Reader) (*api.MsgGenericKprobeArgError, error) {
 	}
 
 	if status != 0 {
-		arg.Message = strconv.FormatUint(uint64(status), 10)
+		if status == ^uint32(0) {
```
Comment (kkourt) [inline pkg/sensors/tracing/args_linux.go:115]: Is this the return from `probe_read_str`? Or we have a convention for status?

**Review by kkourt — COMMENTED** [review]
Thanks Andy!

I'm having some difficulty following this PR. I'm not super familiar with the code, so maybe I'm missing something obvious. Could you please add some explanations to help me understand the change (e.g., iin commit messages)?

File: bpf/process/types/basic.h (hunk 1 above)
Comment (andrewstrohman) [reply bpf/process/types/basic.h:369]: `bpf_probe_read_str` will not return 0 under any condition. See [here](https://docs.ebpf.io/linux/helper-function/bpf_probe_read_str/#:~:text=On%20success%2C%20the%20strictly%20positive,On%20error%2C%20a%20negative%20value.). It either returns a positive number, to indicate success and the number of bytes copieid (including NULL terminator) or it returns a negative number to indicate failure. 

The reason the success case is a positive number (as opposed to 0 being valid) is because it includes the NULL terminator.

It would have been OK to leave the condition as it was, so if you prefer it to remain that way, I'll revert that part. I don't think there is any functional/behavioral difference though.

File: bpf/process/generic_calls.h (hunk 2 above)
Comment (andrewstrohman) [reply bpf/process/generic_calls.h:666]: I removed this because it's dead code now. `generic_read_arg` will no longer return a negative return value under after this change, so I thought removing it would prevent confusion. I have added this information into the commit message.

File: bpf/process/generic_calls.h (hunk 3 above)
Comment (andrewstrohman) [reply bpf/process/generic_calls.h:474]: path_arg is not special. In the future, I'd like `copy_path` to return a negative value if we encounter a fault. If I kept the early return for `if (path_arg)`, then I would need to add the fault detection logic and arg status fixup for that early return too. This way, there is a single point where we evaluate if a fault occurred and update the status accordingly.

I've updated the commit message to describe this thinking.

File: pkg/sensors/tracing/tracepoint_test.go (hunk 4 above)
Comment (andrewstrohman) [reply pkg/sensors/tracing/tracepoint_test.go:482]: Prior to the arg status change, when `generic_read_arg` returned a negative value (ie NULL string), we would `return filter_args_reject(e->func_id)`.  The `return` part was the cause of the event drop. The calling of `filter_args_reject(e->func_id)` made the generic_tracepoint_process_event program reference the retprobe_map.

Now that `generic_read_arg` will no longer return negative values, the dropping issue is fixed, and the program no longer has a connection to the `retprobe_map` map.

I've tried to improve the commit message to explain this.

File: pkg/sensors/tracing/args_linux.go (hunk 5 above)
Comment (andrewstrohman) [reply pkg/sensors/tracing/args_linux.go:115]: With my current approach, on the BPF side, if a negative return code is encountered from `probe_read_str`, we set arg status code to -1, regardless of the reason for error.

We could preserve/plumb the error codes in the arg status if you think that's valuable. We currently don't relay the error code for the resolve dereference error case either. So if we want to plumb the return code so that it can be reflected in the event, I'd like to do the same for resolve induced errors as well.

File: bpf/process/types/basic.h (hunk 1 above)
Comment (kkourt) [reply bpf/process/types/basic.h:369]: If it's not possible to return `0`, no need to change it, thanks!

File: pkg/sensors/tracing/uprobe_test.go [hunk 6]
```diff
@@ -1127,3 +1127,63 @@ spec:
 	err = jsonchecker.JsonTestCheck(t, checker)
 	require.NoError(t, err)
 }
+
+func TestUprobeNULLStringAndReturnArg(t *testing.T) {
```
Comment (kkourt) [inline pkg/sensors/tracing/uprobe_test.go:1221]: Is there a reason that we are not using the new policytest format?

**Review by kkourt — APPROVED** [review]
Just a comment on the test, but otherwise this looks good to me, thanks!

File: bpf/process/user_preload.h [hunk 7]
```diff
@@ -140,7 +140,7 @@ try_preload_arg(int idx, struct preload_arg_data *data)
 {
 	asm volatile("%[idx] &= %1 ;\n"
 		     : [idx] "+r"(idx)
-		     : "i"(MAX_POSSIBLE_ARGS));
+		     : "i"(MAX_POSSIBLE_ARGS_MASK));
```
Comment (olsajiri) [inline bpf/process/user_preload.h:143]: nice, so we never caught it because we always used index 0? I guess we should add more tests for other than zero arg/data

**Review by olsajiri — APPROVED** [review]

File: pkg/sensors/tracing/uprobe_test.go (hunk 6 above)
Comment (andrewstrohman) [reply pkg/sensors/tracing/uprobe_test.go:1221]: Yes, I put the reason in the description of this PR:

> I wanted to make the uprobe test a policytest, but I couldn't because there is no way to determine kfunc availability for policy tests yet.

Do you know of a way that I can do this? If not, I can think about adding support for this detection in future, but I didn't want to gate these fixes on that improvement.

File: bpf/process/user_preload.h (hunk 7 above)
Comment (andrewstrohman) [reply bpf/process/user_preload.h:143]: > so we never caught it because we always used index 0?

Yes, this issue only affected args configured at indexes 2 and 3.

> I guess we should add more tests for other than zero arg/data

I've added additional tests to the end of this series. They check that we can handle string args at any position within the function signature and within any configuration position within the args section of the policy.

Please let me know what you think.

File: pkg/sensors/tracing/uprobe_test.go (hunk 6 above)
Comment (olsajiri) [reply pkg/sensors/tracing/uprobe_test.go:1221]: there's WithSkip where you can do the detection IIUC the problem

Comment (andrewstrohman) [reply pkg/sensors/tracing/uprobe_test.go:1221]: Thanks. I'm aware of WithSkip, but the code within WithSkip is run on the local machine, not the agent that's being tested. We need a way to query the agent's features. `si.AgentInfo.Probes` was relayed from the agent to local, but it does not include the kfunc availablity of the agent.

Comment (kkourt) [reply pkg/sensors/tracing/uprobe_test.go:1221]: OK, thanks! @andrewstrohman can you create an issue for this?

Comment (andrewstrohman) [reply pkg/sensors/tracing/uprobe_test.go:1221]: Yes, [here](https://github.com/cilium/tetragon/issues/4913) is the issue.
